mv: reject user requests by coordinator when a replica is overloaded by MVs

wmitros opened the pull request:
Currently, when a view update backlog of one replica is full, the write is still sent by the coordinator to all replicas. Because of the backlog, the write fails on the replica, causing inconsistency that needs to be fixed by repair. To avoid these inconsistencies, this patch adds a check on the coordinator for overloaded replicas. As a result, a write may be rejected before being sent to any replicas and later retried by the user, when the replica is no longer overloaded.

This patch does not remove the replica write failures, because we still may reach a full backlog when more view updates are generated after the coordinator check is performed and before the write reaches the replica.

Fixes https://github.com/scylladb/scylladb/issues/17426

wmitros commented:
The coordinator check is placed just before creating the write_response_handler, at the same place where we're checking the 
amount of memory used for "in flight hints". Perhaps we could avoid some work by placing it earlier in the flow, but at the very least we need to process the affected partition key and discover the replicas whose view update backlogs we need to check.

wmitros commented:
I marked this as a draft for now to check if this is a good direction and also, I believe this needs more tests.

Currently, I added a simple test case inspired by https://github.com/scylladb/scylladb/pull/16819 - it was not difficult to overload the replica using `DELETE`s causing large view updates but as seen in https://github.com/scylladb/scylla-dtest/issues/4115, it can also be achieved using just large `INSERT`s, which would be probably less complicated.

Assuming we handle errors caused by "too many hints in flight" correctly, this new failure should perform just as well, but we should at least consider if there's anything special concerning hints, repairs, read-repairs, batches or alternator writes where we would need to adjust our behavior. 

The only writes I found that do not check for too many hints are paxos (lwt) writes. For now, I added the check also for that workflow, is there an argument why we shouldn't?

mykaul commented:
What happens if one replica has a bad day? It'll cause all writes to fail, whereas the intent of quorum is to acknowledge the world isn't perfect. I'm not sure this is substantially better than failing on the replica.

wmitros commented:
> What happens if one replica has a bad day? It'll cause all writes to fail, whereas the intent of quorum is to acknowledge the world isn't perfect. I'm not sure this is substantially better than failing on the replica.

Makes sense, perhaps we should still admit the write if the number of overloaded replicas is small enough to achieve the cl without them. It would still be better check it on the coordinator level and delay writing to some replicas instead of doing a write that we know will fail. However, with this approach, the situation on the slow replica may keep getting worse, due to the writes that are getting retried (maybe we could also avoid that using some kind of exponential backoff, but for example hints do not use it AFAIK)
On the other hand, this kind of an overload is quite specific, i.e. it's not actually the replica that becomes slow but it's the remote view updates that it sends that are taking too long. Maybe this makes the problem less likely, but it's still may happen due to local network congestion or a similar issue.
cc @nyh

piodul commented:
@nyh ping - could you take a look at this PR?

nyh commented:
@wmitros is this PR supposed to be Fixes https://github.com/scylladb/scylladb/issues/17426 ?

nyh commented on service/storage_proxy.cc:
```diff
@@ -3210,6 +3222,25 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+        }
+        slogger.trace("Max view update backlog: {} on node {}", max_backlog.relative_size(), max_ep);
+        tracing::trace(tr_state, "Max view update backlog: {} on node {}", max_backlog.relative_size(), max_ep);
+        return max_backlog.relative_size() < 1;
```
Nitpick: If always tracing the max backlog isn't important (is it?), a much simpler implementation would just replace all these 14 lines by just
```c++
for (const auto& ep : targets) {
    if (get_backlog_of(ep).relative_size() >= 1) {
        return false;
    }
}
return true;
```

wmitros replied:
Looks good, I changed it

nyh commented on test/topology_custom/test_mv_large_inserts.py:
```diff
@@ -0,0 +1,61 @@
+    while time.time() < deadline:
+        done = await cql.run_async(f"SELECT COUNT(*) FROM system_distributed.view_build_status WHERE status = 'SUCCESS' ALLOW FILTERING")
+        logger.info(f"Views built: {done[0][0]}")
+        if done[0][0] == node_count * mvs_count:
```
Unless I'm missing something, this test looks very fragile - what if other views were built in the past on other tables? I think you should look for the views of one particular table - not just for all the views in the system.

Yes, I know a lot of topology tests assume it uses a fresh cluster (so in particular this code will work). But I think this (assuming a fresh cluster) is a mistake we'll regret (I already regret it), and moreover, in the future we might (?) have internal materialized views or something.

wmitros replied:
While I'm changing this test anyway, I don't think we'll be able to prepare a (minimal) unit test without a fresh cluster - if we want the test to be quick and repeatable, we need to start with some given size of the backlog (empty) so that we know how much data should we use.
Even if it was possible to make a consistent test that can run in an existing cluster, I'm not sure if additionally preparing a framework for "testing the test" is worth it for a single test.

nyh commented on test/topology_custom/test_mv_large_inserts.py:
```diff
@@ -0,0 +1,61 @@
+@skip_mode('release', "error injections aren't enabled in release mode")
+async def test_large_concurrency(manager: ManagerClient) -> None:
+    node_count = 2
+    await manager.servers_add(node_count, cmdline=['--error-injections-at-startup=view_update_limit'])
```
I can't find this view_update_limit injection. Is it a patch you forgot to send? Or maybe I'm just searching incorrectly?

wmitros replied:
Yes, it's part of https://github.com/scylladb/scylladb/pull/16819. I could send it separately but I hope https://github.com/scylladb/scylladb/pull/16819 will get merged soon

nyh commented on test/topology_custom/test_mv_large_inserts.py:
```diff
@@ -0,0 +1,61 @@
+                    "WHERE c IS NOT NULL and key IS NOT NULL PRIMARY KEY (c, key) ")
+    await wait_for_views(cql, 1, node_count)
+    with pytest.raises(Exception, match="View update backlog is too high"):
+        await insert_with_concurrency(cql, 1000)
```
Interesting, the CQL driver doesn't retry a query automatically if it fails because of overload?

wmitros replied:
I looks like it does, so I'm not sure how this test was passing, but it wasn't consistent for me so I'm changing the retry policy in the new test anyway

mykaul commented:
> > What happens if one replica has a bad day? It'll cause all writes to fail, whereas the intent of quorum is to acknowledge the world isn't perfect. I'm not sure this is substantially better than failing on the replica.
> 
> It depends what you mean by "replica" and "bad day":
> 
> 1. If a **view** replica is **down**, the base replica will know it and will write a view hint and will _not_ count these view hints as "view update backlog".

And I guess it won't be counted towards a 'quorum write'. I did not refer to this situation, but it's good to mention it.

> 2. If a **view** replica is unusually slow, the base replica will increase the view update backlog, and if it continues to increase, the flow control will delay the client more and more. So a view replica with a slow day will reduce performance. This is a pre-existing condition (I wouldn't even call it a problem). What is being added here is that if the client doesn't slow down (or if it has such high concurrency that 1-second sleeps aren't enough), requests will start failing. I think this is fine - I think it's acceptible for requests to temporarily fail during congestion, and the client retries. This is normal load shedding.

This is where I'm unsure yet if it makes things better or worse than before. If we have RF=3 and CL=quorum, and one replica is slow, that will now fail, whereas before it would have succeeded, no? Would a hint be written, btw?

wmitros commented on service/storage_proxy.cc:
```diff
@@ -3210,6 +3222,25 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
 }
 
+template<typename Range>
+bool storage_proxy::can_update_views(const Range& targets, const schema_ptr& s, tracing::trace_state_ptr tr_state) const {
```
Added a comment

wmitros commented:
The testing of this issue is complicated by https://github.com/scylladb/scylladb/issues/18461,  https://github.com/scylladb/scylladb/issues/18462 and https://github.com/scylladb/scylladb/issues/18542, but I added a new test anyway because I think it's better than what I uploaded initially.

Also, after discussing with @piodul, we decided that it makes more sense to make the admission control more strict than the view backlog based mechanism for rejecting replica writes, particularly because currently, if we ever reach a full view update backlog, the mechanism will reject the replica write and only after the admission control will have a chance to reject writes.
The higher restriction is not admitting mutations when a view update backlog is close to filling up (I set it arbitrarily as 80% for now) instead of when it's already full.

wmitros commented:
This patch only entails the last 3 commits, the previous 2 are https://github.com/scylladb/scylladb/pull/19341 which is needed for it to work

wmitros commented:
> > 2. If a **view** replica is unusually slow, the base replica will increase the view update backlog, and if it continues to increase, the flow control will delay the client more and more. So a view replica with a slow day will reduce performance. This is a pre-existing condition (I wouldn't even call it a problem). What is being added here is that if the client doesn't slow down (or if it has such high concurrency that 1-second sleeps aren't enough), requests will start failing. I think this is fine - I think it's acceptible for requests to temporarily fail during congestion, and the client retries. This is normal load shedding.
> 
> This is where I'm unsure yet if it makes things better or worse than before. If we have RF=3 and CL=quorum, and one replica is slow, that will now fail, whereas before it would have succeeded, no? Would a hint be written, btw?

Sorry for not replying immediately, this patch was blocked by many issues with the view update backlogs but almost all are fixed now.

In the updated version I kept the check for all replicas that we're writing to, particularly because we would in fact keep writing hints to the slow replica. If we did the check only for enough replicas to reach QUORUM, eventually, we would also need to stop admitting writes at some point, just later, when the situation is even worse because we have too many hints

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9838/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9838/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_admission_control
:small_blue_diamond: topology_custom/test_mv_backlog
#### Failed Tests (6/120):
- [test_mv_retried_writes_reach_all_replicas.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9838/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.10)
- [test_mv_retried_writes_reach_all_replicas.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9838/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.3)
- [test_mv_retried_writes_reach_all_replicas.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9838/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.4)
- [test_mv_retried_writes_reach_all_replicas.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9838/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.7)
- [test_mv_retried_writes_reach_all_replicas.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9838/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.8)
- [test_mv_retried_writes_reach_all_replicas.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9838/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.9)
#### Build Details:
- Duration: 3 hr 20 min
- Builder: i-09b3ff27900b0c074 (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9869/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9869/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_admission_control
:small_blue_diamond: topology_custom/test_mv_backlog
#### Failed Tests (2/120):
- [test_mv_retried_writes_reach_all_replicas.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9869/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.10)
- [test_mv_retried_writes_reach_all_replicas.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9869/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.3)
#### Build Details:
- Duration: 52 min
- Builder: spider8.cloudius-systems.com

wmitros commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9869/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9869/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_custom/test_mv_admission_control 🔹 topology_custom/test_mv_backlog
> 
> #### Failed Tests (2/120):
> * [test_mv_retried_writes_reach_all_replicas.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9869/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_10) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.10)
> * [test_mv_retried_writes_reach_all_replicas.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9869/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.3)
> 
> #### Build Details:
> * Duration: 52 min
> * Builder: spider8.cloudius-systems.com

Looks like we can still reach the race from https://github.com/scylladb/scylladb/pull/18334#issuecomment-2181576939.
In that case, I modified the test so that we don't send multiple requests at the same time at all, due to a delay that is added between user requests. During the delay we should wait for the replica write to the slow node even after reaching quorum and that should keep the backlog up-to-date.

wmitros commented:
@piodul @nyh @avikivity please take a look at this after https://github.com/scylladb/scylladb/pull/19341

nyh commented:
> @piodul @nyh @avikivity please take a look at this after #19341

If I understand correctly that #19341 needs to go in first, can you please mark this one as a draft?

wmitros commented on service/storage_proxy.cc:
```diff
@@ -3208,6 +3208,12 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
+        // A base replica won't be able to send its view updates because there's too many of them.
+        // To avoid any base or view inconsistencies, the request should be retried when the replica
+        // isn't overloaded
+        throw overloaded_exception("View update backlog is too high on node containing the base replica");
```
I modified the throw into a return. Unfortunately, we couldn't just use implicit conversion here - I explained why in an additional commit

avikivity replied:
It still looks like a throw.

wmitros commented on test/topology_custom/test_mv_admission_control.py:
```diff
@@ -0,0 +1,113 @@
+    i = 0
+    while local_node == get_replicas(cql, i)[0]:
+        i = i + 1
+    return i
```
I removed almost all of these when modifying the test for tablets

wmitros commented on test/topology_custom/test_mv_admission_control.py:
```diff
@@ -0,0 +1,113 @@
+    raise Exception("Timeout waiting for views to build")
+
+def get_replicas(cql, key):
+    return cql.cluster.metadata.get_replicas("ks", Int32Type.serialize(key, cql.cluster.protocol_version))
```
I modified the test to use tablets in the rebase

wmitros commented on test/topology_custom/test_mv_admission_control.py:
```diff
@@ -0,0 +1,113 @@
+        # without updating the view backlog in between.
+        await asyncio.sleep(0.2)
+
+    await manager.server_stop(servers[0].server_id)
```
The test is quite large already so I didn't add it in the rebase yet. Also, maybe this should be a separate test case?

wmitros commented:
https://github.com/scylladb/scylladb/pull/19341 got merged so we can continue with this @piodul @nyh @avikivity. In the new version I fixed the last batch of comments and rebased on master

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,25 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+// that the replica is already sending is too high, the replica should not send any more view updates, and consequently,
+// we should not send the write to the replica.
+template<typename Range>
+bool storage_proxy::can_update_views(const Range& targets, const schema_ptr& s, db::write_type type, tracing::trace_state_ptr tr_state) const {
```
The template here is unfortunate. It causes bloat and reduces performance.

I think the main problem is the use of boost::range::join(). It's also pretty bad as it generates complex iterators that then bloat everything that touches them. It's better to copy into a small_vector and then use std::span<> to communicate with other functions.

However, it's all a preexisting problem. We can ignore it for now but please do a cleanup pass later that removes the boost::range::join and the template.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,25 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+    if (type == db::write_type::VIEW) {
+        return true;
+    }
+    if (!_db.local().find_column_family(s).views().empty() && targets.begin() != targets.end()) {
```
This is also unfortunate. We have to do a hash lookup only to find that there are no views. What could have been three instructions for an early return turns out to be many.

Again the fix isn't simple. We can't adjust the schema class since it's supposed to be immutable (except for true schema changes) and adding a view doesn't change the base table. I think we need to add a wrapper for the table+views and look it up instead of the schema, at the top level. /cc @tgrabiec

avikivity replied:
The targets.empty() check is implicit in the for loop.

tgrabiec replied:
We now have `s->table()` so we could avoid the lookup. There was controversy around it, since table is not part of the schema. But `s->views()` seems untroversial since we stay in the schema domain.

wmitros replied:
I used the `s->table()` in the rebase for now.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,25 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+        return true;
+    }
+    if (!_db.local().find_column_family(s).views().empty() && targets.begin() != targets.end()) {
+        for (auto ep : targets) {
```
auto&

wmitros replied:
No longer relevant when using `std::ranges::all_of`

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,25 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+                return false;
+            }
+        }
+    }
```
Could also be written with std::ranges::all_of(targets, ...)

wmitros replied:
I used this in the rebase

avikivity commented on exceptions/coordinator_result.hh:
```diff
@@ -32,7 +32,8 @@ using coordinator_exception_container = utils::exception_container<
     read_failure_exception,
-    rate_limit_exception
+    rate_limit_exception,
+    overloaded_exception
```
ok, since you add the exception here it can be justified to be a separate patch.

piodul replied:
I think it would still be a bit less confusing if `overloaded_exception` were introduced to the list of variants first, and only then new places which return it were introduced (instead of introducing `throw`s first and then changing them to returns).

avikivity commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -46,3 +46,34 @@ async def test_view_backlog_increased_after_write(manager: ManagerClient) -> Non
+    [await manager.api.enable_injection(s.ip_addr, "never_finish_remote_view_updates", one_shot=False) for s in servers]
+    await cql.run_async(stmt, [0, 0, 240000*'a'], host=hosts[0])
+    [await manager.api.disable_injection(s.ip_addr, "never_finish_remote_view_updates") for s in servers]
+    time.sleep(1)
```
Do we really need to waste a second here?

wmitros replied:
We're waiting for the view update backlog gossip round to be performed, but actually, we can just rely on retries instead - I changed this in the rebase

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10267/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10267/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_admission_control
:small_blue_diamond: topology_custom/test_mv_backlog
#### Failed Tests (28/120):
- [test_mv_retried_writes_reach_all_replicas.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.3)
- [test_mv_retried_writes_reach_all_replicas.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.4)
- [test_mv_retried_writes_reach_all_replicas.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.8)
- [test_mv_retried_writes_reach_all_replicas.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_admission_control/test_mv_retried_writes_reach_all_replicas_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_retried_writes_reach_all_replicas.debug.9)
- [test_gossip_same_backlog.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_backlog/test_gossip_same_backlog_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gossip_same_backlog.debug.1)
- [test_gossip_same_backlog.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_backlog/test_gossip_same_backlog_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gossip_same_backlog.debug.10)
- [test_gossip_same_backlog.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_backlog/test_gossip_same_backlog_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gossip_same_backlog.debug.2)
- [test_gossip_same_backlog.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_backlog/test_gossip_same_backlog_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gossip_same_backlog.debug.3)
- [test_gossip_same_backlog.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_backlog/test_gossip_same_backlog_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gossip_same_backlog.debug.4)
- [test_gossip_same_backlog.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10267/testReport/junit/%28root%29/test_mv_backlog/test_gossip_same_backlog_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gossip_same_backlog.debug.5)
#### Build Details:
- Duration: 2 hr 8 min
- Builder: i-0de7d6c3483348eca (m5ad.8xlarge)

wmitros commented on service/storage_proxy.cc:
```diff
@@ -3208,6 +3208,12 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
+        // A base replica won't be able to send its view updates because there's too many of them.
+        // To avoid any base or view inconsistencies, the request should be retried when the replica
+        // isn't overloaded
+        throw overloaded_exception("View update backlog is too high on node containing the base replica");
```
Actually, we check if the table has views in `can_update_views` so the `db::write_type::VIEW` check is probably not strictly necessary. But this check could be faster for view updates than checking if a table has views, so I'm leaving it as is

wmitros replied:
> Actually, we check if the table has views in `can_update_views` so the `db::write_type::VIEW` check is probably not strictly necessary. But this check could be faster for view updates than checking if a table has views, so I'm leaving it as is

When fixing https://github.com/scylladb/scylladb/pull/18334#discussion_r1668782223, I made the check of whether the table has views much less expensive so the `db::write_type::VIEW` check isn't really making a difference anymore so I'm removing it in the next rebase

wmitros commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -46,3 +46,34 @@ async def test_view_backlog_increased_after_write(manager: ManagerClient) -> Non
+async def test_gossip_same_backlog(manager: ManagerClient) -> None:
+    node_count = 2
+    # Use a higher smp to make it more likely that the writes go to a different shard than the coordinator.
+    servers = await manager.servers_add(node_count, cmdline=['--smp', '5'], config={'error_injections_at_startup': ['view_update_limit', 'update_backlog_immediately'], 'enable_tablets': True})
```
AFAIK, we can force a specific tablet layout, but we can't transfer tablets between shards on one node yet (should be doable using 2 migrations). The driver definitely should be able to connect to a specific shard, but I haven't seen any public methods facilitating this.
Fortunately, in this case we don't actually need more shards, the issue reproduces quite consistently even without the higher number, looks like this config was a leftover from some earlier version

wmitros commented:
In the rebase I fixed the last batch of comments, cleaned up the implementation a bit and fixed the CI errors - while doing that I stopped relying on a sleep in the `test_mv_retried_writes_reach_all_replicas` case to avoid any possible flakiness

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10304/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10304/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_admission_control
:small_blue_diamond: topology_custom/test_mv_backlog
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10304/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10551/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2526/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10304/)
#### Build Details:
- Duration: 3 hr 20 min
- Builder: spider2.cloudius-systems.com

wmitros commented:
@avikivity @piodul @nyh can we try continuing with this? All the comments until the last review have been addressed, but the latest version had no reviews yet

piodul commented on service/storage_proxy.cc:
```diff
@@ -3207,14 +3207,14 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
         // a small number of nodes causing problems, so we should avoid shutting down writes completely to
         // healthy nodes.  Any node with no hintsInProgress is considered healthy.
-        throw overloaded_exception(_hints_manager.size_of_hints_in_progress());
+        return boost::outcome_v2::failure(overloaded_exception(_hints_manager.size_of_hints_in_progress()));
```
> because the exception, converted into coordinator_exception_container
> can be then converted into both into a successful response_id_type
> as well as into a failure.

How can `coordinator_exception_container` be converted into `response_id_type` ?? How does that work?

wmitros replied:
`response_id_type` is just an `uint64_t` and it seems like the `coordinator_exception_container`'s bool operator is used for the conversion so we get 1 or 0

nyh commented on test/topology_custom/test_mv_admission_control.py:
```diff
@@ -0,0 +1,132 @@
+
+    # Prepare the statement so that the write goes to the same shard both
+    # times (the first write will cause only the shard on which it was
+    # performed to have the updated view update backlog).
```
This "goes to the same shard" might only work with a specific choice of CQL driver (Scylla's one). Do you really need this assumption? What happens if it isn't true (e.g., I run the test using Cassandra's driver) - will the test fail, or just fail to reproduce the bug, or what?

wmitros replied:
Unfortunately, we need this assumption because the information about the increased backlog will only be received on the shard that sends (and receives response for) the first write. With a different (non-shard aware) driver, the increased view backlog will not be noticed so we won't get the admission control error which we expect. If in the future we'll want to try running this with Cassandra's driver, we'll probably need to either skip this test, propagate the received backlogs to other shards, or find another way to send requests to a specific Scylla shard

nyh commented on test/topology_custom/test_mv_admission_control.py:
```diff
@@ -0,0 +1,132 @@
+    bnd_stmt = BoundStatement(stmt, retry_policy=FallthroughRetryPolicy())
+    await cql.run_async(bnd_stmt.bind([0, 0, 240000*'a']), host=hosts[0])
+    with pytest.raises(Exception, match="View update backlog is too high"):
+        await cql.run_async(bnd_stmt.bind([0, 0, 240000*'a']), host=hosts[0])
```
I didn't understand this. Why is the second write already too much backlog? 240KB doesn't seem so much. Did you configure the backlog to be artificially low? Or is it this low just because it's 80% of 10% of something relatively small?
If the answer to my last question is yes, it means that this test will break if we ever change the default amount of memory or default amount of memory dedicated to view backlog. That isn't nice... I think in that case it will be better to have this test explicitly configure Scylla's memory limit, or, alternatively, have the test inquire Scylla's configuration and calculate this 240000 number from it.

nyh replied:
Oh, or maybe the view_update_limit injection is what caused the second write to fail?
But if that's the case, why did you need that large 240KB write? Wouldn't the string 'a' fail just the same?

wmitros replied:
Right, it's due to the `view_update_limit` injection - it sets the view backlog limit to 250000 bytes. It would probably be clearer to show that just any following write fails (like a write with a single 'a'), so I'm replacing the second write with a tiny one in the rebase

wmitros commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -46,3 +46,33 @@ async def test_view_backlog_increased_after_write(manager: ManagerClient) -> Non
+
+    [await manager.api.enable_injection(s.ip_addr, "never_finish_remote_view_updates", one_shot=False) for s in servers]
+    await cql.run_async(stmt, [0, 0, 240000*'a'], host=hosts[0])
+    [await manager.api.disable_injection(s.ip_addr, "never_finish_remote_view_updates") for s in servers]
```
Ok

wmitros commented on test/topology_custom/test_mv_admission_control.py:
```diff
@@ -0,0 +1,132 @@
+        # Based on whether the response from the slow node is received before the next write,
+        # the following small write can serve two purposes:
+        # 1. If the response is received before the next write, the write will be rejected by
+        #   admission control and retried until it reaches all replicas.
```
Right, the driver does the retry. Also, because it's an `overloaded` exception, it looks like we actually keep retrying indefinitely as can be seen in https://github.com/scylladb/python-driver/blob/8a4387ae3f36522cc10841c607abb8182c6f8286/cassandra/policies.py#L870-L893

piodul replied:
Are you sure the driver retries indefinitely? As far as I know, in drivers "retry on next host" usually means "retry on the next node in the plan created by the load balancing policy" (plan == ordered list of replicas). I would guess that there will be at most as many retries as there are nodes in the cluster.

wmitros replied:
That's what I observed. This was actually a problem in one of the tests I was trying along the way, when the server wouldn't stop replying with an 'overloaded' message and the test would never finish. Though I'm not sure if we can avoid that problem in the future - it can potentially occur for every `run_async` call when we're not using a custom retry policy

wmitros commented on test/topology_custom/test_mv_backlog.py:
```diff
@@ -46,3 +46,33 @@ async def test_view_backlog_increased_after_write(manager: ManagerClient) -> Non
+    await cql.run_async(stmt, [0, 0, 240000*'a'], host=hosts[0])
+    [await manager.api.disable_injection(s.ip_addr, "never_finish_remote_view_updates") for s in servers]
+    # The next write should be admitted eventually, after a gossip round (1s) is performed
+    await cql.run_async(stmt, [0, 0, 'a'], host=hosts[0])
```
Similarly to above, we'll keep retrying until one second passes

wmitros commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,23 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+bool storage_proxy::can_update_views(const Range& targets, const schema_ptr& s, db::write_type type) const {
+    static constexpr double mv_admission_control_threshold = 0.8;
+    if (type == db::write_type::VIEW) {
+        return true;
```
Your understanding is right. This function should only reject the write when we have a base table write with views, but view write will still execute it so this check was for returning quickly when we know the write doesn't generate view updates.
However, I see that it's redundant with the following `if` which is also just a few pointer dereferences so I think I'll remove this.

wmitros commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,23 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+// we should not send the write to the replica.
+template<typename Range>
+bool storage_proxy::can_update_views(const Range& targets, const schema_ptr& s, db::write_type type) const {
+    static constexpr double mv_admission_control_threshold = 0.8;
```
It's a fraction of that other threshold (which is also just the view backlog limit). It could be useful to make both of them configurable (maybe for some tests), but I'm not sure we can actually select "optimal" values for them and make them useful in practice.
Note that as @piodul mentioned in https://github.com/scylladb/scylladb/pull/18334#discussion_r1695412481, using 0.8 here effectively lowers the max mv flow control delay to 512ms (if we were to use a longer delay we will now reject the write). If we want to make this configurable we might want to consider making that delay configurable first (I guess we have https://github.com/scylladb/scylladb/issues/18187 for that)

nyh replied:
Right, https://github.com/scylladb/scylladb/issues/18187 says that 1s has this special magical meaning that we need to reconsider, and this patch changes it to 0.5 seconds for no obvious reason. Maybe the code here should use 1.0 and not 0.8, and then to avoid the per-replica code accidentally kicking in before this one does, change the >1 in that code to >1.1 or something?

wmitros replied:
Ok, I tried to achieve something of the sort in the rebase: now we reach the 1s delay on the coordinator when we approach the admission control limit. At the same time we still reject replica writes only when we use all `max_memory_pending_view_updates()` bytes for the backlog

wmitros commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,23 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+    if (type == db::write_type::VIEW) {
+        return true;
+    }
+    if (!s->table().views().empty()) {
```
I used the suggestion in the rebase

wmitros commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,23 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
+            return get_backlog_of(ep).relative_size() < mv_admission_control_threshold;
+        });
+    }
+    return true;
```
I used a more specific name in the rebase that should be clear even if there's no views. I'm not sure if it's that much better so please take a look.

wmitros commented on service/storage_proxy.cc:
```diff
@@ -3296,6 +3308,23 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
 
+// After sending the write to replicas(targets), the replicas might generate send view updates. If the number of view updates
+// that the replica is already sending is too high, the replica should not send any more view updates, and consequently,
+// we should not send the write to the replica.
```
I expanded the comment a little in the rebase

wmitros commented on db/view/view.cc:
```diff
@@ -2652,7 +2652,7 @@ void node_update_backlog::add(update_backlog backlog) {
 update_backlog node_update_backlog::fetch() {
     auto now = clock::now();
-    if (now >= _last_update.load(std::memory_order_relaxed) + _interval) {
+    if (utils::get_local_injector().enter("update_backlog_immediately") || now >= _last_update.load(std::memory_order_relaxed) + _interval) {
```
I added the injection to the constructor in the rebase.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10609/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10609/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_admission_control
:small_blue_diamond: topology_custom/test_mv_backlog
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10609/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10815/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2793/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10609/)
#### Build Details:
- Duration: 3 hr 22 min
- Builder: spider2.cloudius-systems.com

piodul commented:
@wmitros if the PR is ready for a re-review, please write down a short changelog and re-request review via the GitHub interface from the relevant people

wmitros commented:
In the last rebases I fixed all of the comments which includes:
* reworking the method for checking whether we should stop admitting writes due to view backlog
* cleaning up the tests
* keeping the max 1s delay before the admission control kicks while the `max_memory_pending_view_updates()` (10% of memory) is still the limit before rejecting replica writes

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10705/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10705/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_admission_control
:small_blue_diamond: topology_custom/test_mv_backlog
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10705/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10893/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2871/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10705/)
#### Build Details:
- Duration: 3 hr 26 min
- Builder: spider4.cloudius-systems.com

mykaul commented:
@wmitros - please resolve all conversations you've responded to in follow-up patches (where the changelog should reflect those comments). There are too many open ones atm.

piodul commented:
Queued
